internal/encryption/encryption.go

```diff
@@ -41,34 +42,27 @@ func New(reg registry.Registry, cfg *config.Config) Encryption {
 	}
 }
 
-func (e *encryption) StateFile() StateEncryption {
-	return &stateEncryption{
-		base: newBaseEncryption(e, e.cfg.StateFile.AsTargetConfig(), e.cfg.StateFile.Enforced, "statefile"),
-	}
+func (e *encryption) StateFile() (StateEncryption, hcl.Diagnostics) {
+	return newStateEncryption(e, e.cfg.StateFile.AsTargetConfig(), e.cfg.StateFile.Enforced, "statefile")
 }
 
-func (e *encryption) PlanFile() PlanEncryption {
-	return &planEncryption{
-		base: newBaseEncryption(e, e.cfg.PlanFile.AsTargetConfig(), e.cfg.PlanFile.Enforced, "planfile"),
-	}
+func (e *encryption) PlanFile() (PlanEncryption, hcl.Diagnostics) {
+	return newPlanEncryption(e, e.cfg.PlanFile.AsTargetConfig(), e.cfg.PlanFile.Enforced, "planfile")
 }
 
-func (e *encryption) Backend() StateEncryption {
-	return &stateEncryption{
-		base: newBaseEncryption(e, e.cfg.StateFile.AsTargetConfig(), e.cfg.StateFile.Enforced, "backend"),
-	}
+func (e *encryption) Backend() (StateEncryption, hcl.Diagnostics) {
+	return newStateEncryption(e, e.cfg.StateFile.AsTargetConfig(), e.cfg.StateFile.Enforced, "backend")
 }
 
-func (e *encryption) RemoteState(name string) ReadOnlyStateEncryption {
+func (e *encryption) RemoteState(name string) (ReadOnlyStateEncryption, hcl.Diagnostics) {
 	for _, remoteTarget := range e.cfg.Remote.Targets {
 		if remoteTarget.Name == name {
-			return &stateEncryption{
-				// TODO the addr here should be generated in one place.
-				base: newBaseEncryption(e, remoteTarget.AsTargetConfig(), false, "remote.remote_state_datasource."+remoteTarget.Name),
-			}
+			// TODO the addr here should be generated in one place.
+			addr := "remote.remote_state_datasource." + remoteTarget.Name
```

Comment:
This address is weird to me. Isn't it called `datasource.terraform_remote_state`?

Or is this meant to be the address of the encryption method?

I'm a bit confused here.

Reply:
This is the address of the remote encryption configuration for use in errors
`terraform { encryption{ remote { remote_data_source "foo" {} } } } `

We may want to rename that in the future.